include/alp/imf.hpp

```diff
@@ -50,107 +50,202 @@
-                        && s == dynamic_cast< const Strided & >( other ).s;
-                }
-        };
+				const size_t n;
+				const size_t N;
+
+				IMF( const size_t n, const size_t N ): n( n ), N( N ) {}
 
-        class Select: public IMF {
+			protected:
+
+				size_t map( const size_t i ) const {
+					(void)i;
+					std::cout << "Calling IMF map" << std::endl;
+					return 0;
+				}
 
-            public:
-                std::vector< size_t > select;
+				template< typename OtherImf >
+				bool isSame( const OtherImf & other ) const {
+					//static_assert( std::is_same< decltype( *this ), decltype( other ) >::value );
+					return n == other.n && N == other.N;
+				}
+
+		}; // class IMF
 
-                size_t map(size_t i) {
-                    return select.at( i );
-                }
+		/**
+		 * The strided IMF.
+		 * \f$I_n =[0, n), I_N =[0, N)\f$
+		 * \f$Strided_{b, s} = I_n \rightarrow I_N; i \mapsto b + si\f$
+		 */
 
-                Select(size_t N, std::vector< size_t > & select): IMF( select.size(), N ), select( select ) {
-                    if ( *std::max_element( select.cbegin(), select.cend() ) >= N) {
-                        throw std::runtime_error("IMF Select beyond range.");
-                    }
-                }
+		class Strided: public IMF {
 
-                virtual bool isSame( const IMF & other ) const {
-                    return IMF::isSame( other )
-                        && select == dynamic_cast< const Select & >( other ).select;
-                }
-        };
+			public:
+				const size_t b;
+				const size_t s;
 
-        /**
-         * A composition of two IMFs.
-         * \f$I_{g,n} =[0, n), I_{g,N} =[0, N)\f$
-         * \f$I_{f,n} =[0, n), I_{f,N} =[0, N)\f$
-         * \f$Composed_{f, g} = I_{g,n} \rightarrow I_{f,N}; i \mapsto f( g( i ) )\f$
-         */
-        
-        class Composed: public IMF {
+				size_t map( const size_t i ) const {
+					std::cout << "Calling Strided map" << std::endl;
+					return b + s * i;
+				}
 
-            public:
-                std::shared_ptr<IMF> f, g;
+				Strided( const size_t n, const size_t N, const size_t b, const size_t s ): IMF( n, N ), b( b ), s( s ) { }
+
+				template< typename OtherIMF >
+				bool isSame( const OtherIMF &other ) const {
+					return IMF::isSame( other ) &&
+						b == static_cast< const Strided & >( other ).b &&
+						s == static_cast< const Strided & >( other ).s;
+				}
+		};
+
+		/**
+		 * The identity IMF.
+		 * \f$I_n = [0, n)\f$
+		 * \f$Id = I_n \rightarrow I_n; i \mapsto i\f$
+		 */
+
+		class Id: public Strided {
+
+			public:
 
-                size_t map(size_t i) {
-                    return f->map( g->map( i ) );
-                }
+				Id( const size_t n ) : Strided( n, n, 0, 1 ) {}
+		};
 
-                Composed(std::shared_ptr<IMF> f, std::shared_ptr<IMF> g): IMF( g->n, f->N ), f(f), g(g) { }
+		class Select: public IMF {
+
+			public:
 
-        };
+				std::vector< size_t > select;
```

Comment:
If this IMF is selecting a couple of indices based on a given vector of indices, should it not refer to an ALP vector?